File: js/packages/mechanisms/casper/src/exact/client/scheme.ts
import { PaymentPayloadResult, PaymentRequirements, SchemeNetworkClient } from "@x402/core/types";
import { buildDomain, CASPER_DOMAIN_TYPES, hashTypedData } from "@casper-ecosystem/casper-eip-712";
import { SCHEME_EXACT } from "../../constants";
import { ClientCasperSigner } from "../../signer";
import { isValidAddress, isValidContractPackageHash } from "../../utils";

const transferWithAuthorizationTypes = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" },
  ],
};

/**
 * Encode a byte array as a lowercase hex string.
 *
 * @param bytes - The bytes to encode.
 * @returns Hex string without "0x" prefix.
 */
function hexEncode(bytes: Uint8Array): string {
  return Array.from(bytes, b => b.toString(16).padStart(2, "0")).join("");
}

/**
 * Casper client implementation for the Exact payment scheme.
 */
export class ExactCasperScheme implements SchemeNetworkClient {
  readonly scheme = SCHEME_EXACT;

  /**
   * Create a new ExactCasperScheme instance.
   *
   * @param signer - The Casper signer for the payer.
   */
  constructor(private readonly signer: ClientCasperSigner) {}

  /**
   * Create a payment payload for the Exact scheme.
   *
   * @param x402Version - The x402 protocol version.
   * @param paymentRequirements - The payment requirements selected by the client.
   * @returns Promise resolving to the payment payload result.
   */
  async createPaymentPayload(
    x402Version: number,
    paymentRequirements: PaymentRequirements,
  ): Promise<PaymentPayloadResult> {
    if (!isValidContractPackageHash(paymentRequirements.asset)) {
      throw new Error(`invalid_exact_casper_client_invalid_asset: ${paymentRequirements.asset}`);
    }

    if (!isValidAddress(paymentRequirements.payTo)) {
      throw new Error(`invalid_exact_casper_client_invalid_pay_to: ${paymentRequirements.payTo}`);
    }

    const name = paymentRequirements.extra?.name;
    const version = paymentRequirements.extra?.version;
    if (typeof name !== "string" || name === "") {
      throw new Error("invalid_exact_casper_client_missing_token_name");
    }
    if (typeof version !== "string" || version === "") {
      throw new Error("invalid_exact_casper_client_missing_token_version");
    }

    const domain = buildDomain(
      name,
      version,
      paymentRequirements.network,
      "0x" + paymentRequirements.asset,
    );

    const now = Math.floor(Date.now() / 1000);
    const validAfter = now - 600;
    const validBefore = now + paymentRequirements.maxTimeoutSeconds;

    const nonce = crypto.getRandomValues(new Uint8Array(32));

    const message = {
      from: "0x" + this.signer.accountAddress(),
      to: "0x" + paymentRequirements.payTo,
      value: BigInt(paymentRequirements.amount),
      validAfter: BigInt(validAfter),
      validBefore: BigInt(validBefore),
      nonce: "0x" + hexEncode(nonce),
    };

    let digest: Uint8Array;
    try {
      digest = hashTypedData(
        domain,
        transferWithAuthorizationTypes,
        "TransferWithAuthorization",
        message,
        { domainTypes: CASPER_DOMAIN_TYPES },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`invalid_exact_casper_client_failed_to_hash: ${message}`);
    }

    let signature: Uint8Array;
    try {
      signature = await this.signer.signEIP712(digest);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`invalid_exact_casper_client_failed_to_sign: ${message}`);
    }

    return {
      x402Version,
      payload: {
        signature: hexEncode(signature),
        publicKey: this.signer.publicKey(),
        authorization: {
          from: this.signer.accountAddress(),
          to: paymentRequirements.payTo,
          value: paymentRequirements.amount,
          validAfter: String(validAfter),
          validBefore: String(validBefore),
          nonce: hexEncode(nonce),
        },
      },
    };
  }
}

File: go/x402/mechanisms/casper/exact/client/scheme.go
package client

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"math/big"
	"time"

	"casper_x402_facilitator/x402/mechanisms/casper"

	eip712 "github.com/casper-ecosystem/casper-eip-712/go"
	"github.com/x402-foundation/x402/go/types"
)

var transferWithAuthorizationTypes = eip712.TypeDefinitions{
	"TransferWithAuthorization": {
		{Name: "from", Type: "address"},
		{Name: "to", Type: "address"},
		{Name: "value", Type: "uint256"},
		{Name: "validAfter", Type: "uint256"},
		{Name: "validBefore", Type: "uint256"},
		{Name: "nonce", Type: "bytes32"},
	},
}

type ExactCasperScheme struct {
	signer casper.ClientCasperSigner
}

func NewExactCasperScheme(signer casper.ClientCasperSigner) *ExactCasperScheme {
	return &ExactCasperScheme{signer: signer}
}

func (c *ExactCasperScheme) Scheme() string {
	return casper.SchemeExact
}

func (c *ExactCasperScheme) CreatePaymentPayload(
	_ context.Context,
	requirements types.PaymentRequirements,
) (types.PaymentPayload, error) {
	if !casper.IsValidContractPackageHash(requirements.Asset) {
		return types.PaymentPayload{}, fmt.Errorf("%s: %s", ErrInvalidAsset, requirements.Asset)
	}
	contractPackageHash, err := casper.DecodeContractPackageHash(requirements.Asset)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("%s: %w", ErrInvalidAsset, err)
	}

	name, ok := requirements.Extra["name"].(string)
	if !ok || name == "" {
		return types.PaymentPayload{}, errors.New(ErrMissingTokenName)
	}
	version, ok := requirements.Extra["version"].(string)
	if !ok || version == "" {
		return types.PaymentPayload{}, errors.New(ErrMissingTokenVersion)
	}

	domain := eip712.BuildDomain(name, version, requirements.Network, contractPackageHash)

	now := time.Now().Unix()
	validAfter := now - 600
	validBefore := now + int64(requirements.MaxTimeoutSeconds)

	var nonce [32]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to generate nonce: %w", err)
	}

	fromAddr, err := eip712.NewAddressFromHex("0x" + c.signer.AccountAddress())
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to parse from address: %w", err)
	}
	toAddr, err := eip712.NewAddressFromHex("0x" + requirements.PayTo)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("failed to parse to address: %w", err)
	}
	value, ok := new(big.Int).SetString(requirements.Amount, 10)
	if !ok {
		return types.PaymentPayload{}, fmt.Errorf("invalid amount: %s", requirements.Amount)
	}

	message := map[string]interface{}{
		"from":        fromAddr,
		"to":          toAddr,
		"value":       value,
		"validAfter":  big.NewInt(validAfter),
		"validBefore": big.NewInt(validBefore),
		"nonce":       nonce,
	}

	digest, err := eip712.HashTypedData(
		domain,
		transferWithAuthorizationTypes,
		"TransferWithAuthorization",
		message,
		&eip712.TypedDataOptions{DomainTypes: eip712.CasperDomainTypes},
	)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("%s: %w", ErrFailedToHash, err)
	}

	sig, err := c.signer.SignEIP712(digest)
	if err != nil {
		return types.PaymentPayload{}, fmt.Errorf("%s: %w", ErrFailedToSign, err)
	}

	payload := &casper.ExactCasperPayload{
		Signature: hex.EncodeToString(sig[:]),
		PublicKey: c.signer.PublicKey(),
		Authorization: casper.ExactCasperAuthorization{
			From:        c.signer.AccountAddress(),
			To:          requirements.PayTo,
			Value:       requirements.Amount,
			ValidAfter:  fmt.Sprintf("%d", validAfter),
			ValidBefore: fmt.Sprintf("%d", validBefore),
			Nonce:       hex.EncodeToString(nonce[:]),
		},
	}

	return types.PaymentPayload{X402Version: 2, Payload: payload.ToMap()}, nil
}
